File: problems/partition_equal_subset_sum.py
# https://leetcode.com/problems/partition-equal-subset-sum
from typing import List

class Solution:

    def canPartition(self, nums: List[int]) -> bool:
        memo = {}
        if sum(nums) % 2 != 0:
            return False
        def can_partition(nums, s_l, s, n):
            s_r = s - s_l
            if s_l == s_r:
                return True

            if n == len(nums):
                return s_l == s_r


            k = tuple((s_l, n+1))
            
            if memo.get(k) is None:
                memo[k] = can_partition(nums, s_l + nums[n], s, n+1) or can_partition(nums, s_l, s, n+1)

            
            return memo[k]
            

        return can_partition(nums, 0, sum(nums), 0)

File: problems/partition_equal_subset_sum.cpp
#include <bits/stdc++.h>

using namespace std;
class Solution {
public:

    bool subset_sum(int target, vector<int>& nums){
        vector<bool> has_sum(target+1, 0);

        has_sum[0] = true;
        for (auto& v: nums){
            for (int r = target; r >= v; r--){
                has_sum[r] = has_sum[r] || has_sum[r-v];
            }
        }

        return has_sum[target];
    }
    bool canPartition(vector<int>& nums) {
        int sum = 0;

        for (auto& el : nums){
            sum+= el;
        }

        if (sum % 2)
            return false;

        return subset_sum(sum/2, nums);
    }
};